Add extension point in update function to allow additional checks, etc.

hslusarek opened the pull request:
As user user of the `JointTrajectoryController`, I need to be able to verify/check the newly calculated desired state before the [hardware interface  is updated](https://github.com/ros-controls/ros_controllers/blob/melodic-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L471), in order, to react to e.g. Cartesian velocity violations, etc.

This PR tries to continue and, more importantly, tries to minimize the necessary changes needed to achieve the objective of PR #442.

This PR is tightly coupled with the needs of [PR #308 on PilzDE/pilz_robots](https://github.com/PilzDE/pilz_robots/pull/308).

**Please note:**
The general refactoring of the update function is now in a separate PR: #450.

martiniil commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:
```diff
@@ -811,6 +818,23 @@ setHoldPosition(const ros::Time& time, RealtimeGoalHandlePtr gh)
 
+template <class SegmentImpl, class HardwareInterface>
+inline bool JointTrajectoryController<SegmentImpl, HardwareInterface>::
+checkStates(const std::vector<double>& /* old_desired_position */,
```
Do you have a reason for considering only position checks? (besides that it fits our needs)
If not, I would pass the whole states, not only the position.

hslusarek replied:
Done.

martiniil commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller.h:
```diff
@@ -246,6 +246,12 @@ class JointTrajectoryController : public controller_interface::Controller<Hardwa
   void setHoldPosition(const ros::Time& time, RealtimeGoalHandlePtr gh=RealtimeGoalHandlePtr());
 
+private:
+  virtual bool checkStates(const std::vector<double>& old_desired_position,
```
I think it is important to document that these functions have to be "realtime-safe" since they are called from `update()`.

hslusarek replied:
Done.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+///////////////////////////////////////////////////////////////////////////////
+// Copyright (C) 2013, PAL Robotics S.L.
```
Are this test not from us? So shouldn't we the ones with the Copyright here?

martiniil replied:
You are right. But having a whole group of copyright holders in one project can lead to difficulties. So commonly the contributors of a project agree on assigning the copyright to a single person/entity.
@bmagyar What is your general approach for external contributions concerning the copyright?

bmagyar replied:
If this file is entirely your contribution, please attribute it to yourselves. Should be the same bsd with the holder being different

martiniil replied:
Thanks, I adapted the copyright headers of the tests.

hslusarek commented on joint_trajectory_controller/test/stop_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,222 @@
+///////////////////////////////////////////////////////////////////////////////
+// Copyright (C) 2013, PAL Robotics S.L.
```
Are this test not from us? So shouldn't we the ones with the Copyright here?

bmagyar replied:
credit goes to you guys

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+//   * Redistributions in binary form must reproduce the above copyright
+//     notice, this list of conditions and the following disclaimer in the
+//     documentation and/or other materials provided with the distribution.
+//   * Neither the name of PAL Robotics S.L. nor the names of its
```
Adapt if copyright is changed.

bmagyar replied:
credit goes to you guys

hslusarek commented on joint_trajectory_controller/test/stop_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,222 @@
+//   * Redistributions in binary form must reproduce the above copyright
+//     notice, this list of conditions and the following disclaimer in the
+//     documentation and/or other materials provided with the distribution.
+//   * Neither the name of PAL Robotics S.L. nor the names of its
```
Adapt if copyright is changed.

bmagyar replied:
credit goes to you guys

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+class HoldTrajectoryBuilderTest : public testing::Test
+{
+public:
+  void initDefaultTrajectory(unsigned int number_of_joints, Trajectory& trajectory)
```
Actually a `static` function.

martiniil replied:
Moved to separate header. Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/hold_trajectory_builder.h:
```diff
@@ -0,0 +1,110 @@
+  using JointHandle = typename HardwareInterface::ResourceHandleType;
+
+public:
+  HoldTrajectoryBuilder(const unsigned int& number_of_joints,
```
`number_of_joints` probably unnecessary because it can be determined from the `joints` vector.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/hold_trajectory_builder.h:
```diff
@@ -0,0 +1,110 @@
+    hold_start_state_.velocity[0]     =  0.0;
+    hold_start_state_.acceleration[0] =  0.0;
+
+    Segment& segment {(*hold_traj)[joint_index].front()};
```
There are checks missing here: 
- Who guarantees that the `hold_traj` actually contains a `TrajectoryPerJoint` for each needed joint?
- Who guarantees that the `TrajectoryPerJoint` actually contains a segment?

martiniil replied:
I think the checks were left out intentionally, because you guarantee that by construction via `createHoldTrajectory()` in the controller.
However I would not mind adding the checks. What do you think?
Also: How about moving `createHoldTrajectory()` into `hold_trajectory_builder.h`?

hslusarek replied:
>However I would not mind adding the checks. What do you think?

Please, add the checks.

> Also: How about moving createHoldTrajectory() into hold_trajectory_builder.h?

For me the `createHoldTrajectory()` function is like a factory function. If you move the function to the `hold_trajectory_builder.h`, please separate this function from the actual builder. You could add it e.g. as a stand-alone static function.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+    expected_state.position[0] = positions[i];
+
+    trajectory[i][0].sample(trajectory[i][0].startTime(), sampled_state);
+    EXPECT_TRUE(statesAlmostEqual<Segment>(sampled_state, expected_state));
```
In case of a failed check, state for which joint it failed. Additionally, state if the check failed for the start state or for the end state.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+
+  EXPECT_TRUE(builder.buildTrajectory(&trajectory));
+
+  EXPECT_EQ(rt_goal_handle.use_count(), 3);
```
For more test stability: Please calculate from joint number.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+int main(int argc, char** argv)
+{
+  testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
```
I'm missing tests which check what happens if I insert an incomplete trajectory into the `buildTrajectory()` function.

martiniil replied:
Depends on how we proceed with the following:

> There are checks missing here:
> 
>     * Who guarantees that the `hold_traj` actually contains a `TrajectoryPerJoint` for each needed joint?
> 
>     * Who guarantees that the `TrajectoryPerJoint` actually contains a segment?

hslusarek replied:
See my answer above.

martiniil replied:
Done.

hslusarek replied:
Are there tests which test the newly introduced integrity-check function?

martiniil replied:
Only indirectly via the derived classes.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+  EXPECT_EQ(trajectory.size(), number_of_joints);
+  for (const auto& jt : trajectory)
+  {
+    EXPECT_EQ(jt.size(), 1U);
```
Please consequently use a custom failure message via `<<` operator to simplify debugging. See also [googletest/primer.md](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#assertions).

**Please note:**
This comment holds for all Assertions, etc.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+  using Builder = joint_trajectory_controller::HoldTrajectoryBuilder<QuinticSplineSegment, TypeParam>;
+
+  std::vector<typename TypeParam::ResourceHandleType> joints = this->getJointHandles();
+  auto number_of_joints = joints.size();
```
Use `const` if the variable is not supposed to change.

**Please note:**
This comment holds for all not changing variables or members, etc.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+  {
+    EXPECT_EQ(jt.size(), 1U);
+  }
+  EXPECT_NEAR(trajectory[0][0].startTime(), start_time, EPS);
```
I suggest to use `.at(0)` instead of `[0]` because it provides additional safety and in case of an error more detailed error information while the extra computation time is more or less irrelevant for the test.

**Please note:**
This comment holds for all container access operations.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+  this->initDefaultTrajectory(number_of_joints, trajectory);
+
+  std::vector<double> positions{-48.9, 2.1};  // set some arbitrary positions
+  this->robot_.setPositions(positions[0], positions[1]);
```
Please, for OOP consistency add getter method to gain access to the robot.

By the way: the `protected` modifier instead of the `public` modifier should suffice for methods and members.

martiniil replied:
See below.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+ * @note A non-empty goal handle cannot be created without an action server,
+ * therefore we check only how the use_count of the shared pointer changes.
+ */
+TYPED_TEST(HoldTrajectoryBuilderTest, testResetGoalHandle)
```
Please create separate test class for `TrajectoryBuilder` base class, to test functionality of base class. This avoids the code duplication in `hold_trajectory_builder_test.cpp` and `stop_trajectory_builder_test.cpp`. Maybe use GMock to provide implementation for unimplemented functions of base class.

martiniil replied:
See below.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+ * @note A non-empty goal handle cannot be created without an action server,
+ * therefore we check only how the use_count of the shared pointer changes.
+ */
+TYPED_TEST(HoldTrajectoryBuilderTest, testSetGoalHandle)
```
Please create separate test class for `TrajectoryBuilder` base class, to test functionality of base class. This avoids the code duplication in `hold_trajectory_builder_test.cpp` and `stop_trajectory_builder_test.cpp`. Maybe use GMock to provide implementation for unimplemented functions of base class.

martiniil replied:
Thanks for the idea. I added a test for the base class. However, in order to test the effect of `reset()`, `setStartTime()` and `setGoalHandle()` on the outcome of `buildTrajectory()` I cannot avoid the code duplication.

hslusarek replied:
Ok

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+  }
+}
+
+TYPED_TEST(HoldTrajectoryBuilderTest, testResetStartTime)
```
Please create separate test class for `TrajectoryBuilder` base class, to test functionality of base class. This avoids the code duplication in `hold_trajectory_builder_test.cpp` and `stop_trajectory_builder_test.cpp`. Maybe use GMock to provide implementation for unimplemented functions of base class.

martiniil replied:
See above.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+
+private:
+	double pos1_, vel1_, eff1_, cmd1_, vel_cmd1_, acc_cmd1_ = 0.0;
+	double pos2_, vel2_, eff2_, cmd2_, vel_cmd2_, acc_cmd2_ = 0.0;
```
Be careful! Only `acc_cmd1_` and `acc_cmd2_` are set to `0.0`. Write e.g.
```
double pos1_{0}, vel1_{0}, eff1_{0}, cmd1_{0}, vel_cmd1_{0}, acc_cmd1_{0};
```

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/hold_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,324 @@
+{
+static constexpr double EPS{1e-9};
+
+template<class SegmentType>
+bool statesAlmostEqual(const typename SegmentType::State& state1,
+                       const typename SegmentType::State& state2,
+                       const double& tolerance=EPS)
+{
+  using namespace joint_trajectory_controller_tests;
+  return vectorsAlmostEqual(state1.position, state2.position, tolerance) &&
+         vectorsAlmostEqual(state1.velocity, state2.velocity, tolerance) &&
+         vectorsAlmostEqual(state1.acceleration, state2.acceleration, tolerance);
+}
```
Code duplication in `stop_trajectory_builder_test.cpp`.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/stop_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,222 @@
+	  TrajectoryPerJoint joint_traj{segment};
+    trajectory.resize(number_of_joints, joint_traj);
+  }
+};
```
Code duplication with `hold_trajectory_builder_test.cpp`

Function is actually `static`, so maybe move it out of class and into separate header?

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/test/stop_trajectory_builder_test.cpp:
```diff
@@ -0,0 +1,222 @@
+	hold_state.position[0] = 0.9;  // set arbitrary state
+	hold_state.velocity[0] = -0.03;
+	hold_state.position[1] = 1.68;
+	hold_state.velocity[1] = 3.7;
```
I suggest to use `.at(0)` instead of `[0]` because it provides additional safety and in case of an error more detailed error information while the extra computation time is more or less irrelevant for the test.

**Please note:**
This comment holds for all container access operations.

martiniil replied:
Done.

hslusarek reviewed (commented):
Review for `hold_trajectory_builder_test.cpp` and `stop_trajectory_builder_test.cpp`

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/trajectory_builder.h:
```diff
@@ -0,0 +1,165 @@
+
+public:
+  TrajectoryBuilder<SegmentImpl>* setStartTime(const typename Segment::Time& start_time);
+  TrajectoryBuilder<SegmentImpl>* setGoalHandle(RealtimeGoalHandlePtr& goal_handle);
```
I think we should add little bit of documentation here. Questions which should be answered are:
- In which unit do the arguments have to be stated?
- For what are the start time and the goal handle used?
- Are these values ever reset?

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/trajectory_builder.h:
```diff
@@ -0,0 +1,165 @@
+
+  /**
+   * @return True if the trajectory contains a TrajectoryPerJoint for each joint and if it contains the necessary
+   *         number of segments for each TrajectoryPerJoint, otherwise false.
```
- If you rename the function to something like `isTrajectoryValid()` the return value is more intuitive.
- Document the arguments in more detail.
- Maybe rename the arguments `number_of_joints` and `number_of_segments` to `expected_number_of_joints` and `expected_number_of_segments` for more clarity.
- Make the arguments `const`.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/trajectory_builder.h:
```diff
@@ -0,0 +1,165 @@
+{
+
+/**
+ * @brief Base class for classes used to construct diffent trajectory types.
```
Maybe we could add a small example on how to use the builders. Or should this be part of the derived classes? What do you think?

martiniil replied:
I added it to the base class.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/trajectory_builder.h:
```diff
@@ -0,0 +1,165 @@
+
+protected:
+  RealtimeGoalHandlePtr createGoalHandlePtr() const;
+  const boost::optional<typename Segment::Time>& getStartTime() const;
```
Add docu.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/hold_trajectory_builder.h:
```diff
@@ -0,0 +1,111 @@
+  using JointHandle = typename HardwareInterface::ResourceHandleType;
+
+public:
+  HoldTrajectoryBuilder(const std::vector<JointHandle>&  joints);
```
Docu: Maybe describe the argument and state that the reference to the joint handles is stored internally? What do you think?

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/stop_trajectory_builder.h:
```diff
@@ -0,0 +1,145 @@
+public:
+  StopTrajectoryBuilder(const unsigned int& number_of_joints,
+                        const typename Segment::Time& stop_traj_duration,
+                        const typename Segment::State& hold_state);
```
Docu: Explain/Describe the expected arguments and state and state the unit of the arguments (if not obvious). What do you think?

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/stop_trajectory_builder.h:
```diff
@@ -0,0 +1,145 @@
+  const unsigned int number_of_joints_;
+  const typename Segment::Time stop_traj_duration_;
+  // To always have the newest value of the hold state,
+  // a reference is stored.
```
Change the comment into a doxygen documentation and maybe write something like:
"To always have the newest value of the hold state, a reference is stored to the state which has to be reached when the stop-trajectory has to be build via buildTrajecory() function."

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/stop_trajectory_builder.h:
```diff
@@ -0,0 +1,145 @@
+ * @brief Builder creating a trajectory stopping the robot.
+ *
+ * @note The start state of the stop motion and the final settle/hold state
+ * are eqal.
```
Has no relevance anymore and can be deleted.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller.h:
```diff
@@ -246,6 +252,54 @@ class JointTrajectoryController : public controller_interface::Controller<Hardwa
+   *
+   * @param time_data Updated time data.
+   */
+  virtual void updateFuncExtensionPoint(const Trajectory& curr_traj, const TimeData& time_data);
```
Docu: Explain the arguments in more detail and state the unit's if not obvious.

martiniil replied:
Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller.h:
```diff
@@ -246,6 +252,54 @@ class JointTrajectoryController : public controller_interface::Controller<Hardwa
 
+protected:
+  /**
+   * @returns the number of joints of the robot.
```
For the Doxygen Html output it would be better to state what the function does via `@brief` statement instead of using the `@returns` statement. If `@brief` is used, Doxygen displays the explanation what the function does already in the summary at the beginning of the Html page. What do you think?

martiniil replied:
I like `@brief` better. Done.

hslusarek commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller.h:
```diff
@@ -246,6 +252,54 @@ class JointTrajectoryController : public controller_interface::Controller<Hardwa
+   * @returns a trajectory consisting of joint trajectories with one pre-allocated
+   * segment.
+   */
+  static TrajectoryPtr createHoldTrajectory(const unsigned int& number_of_joints);
```
For the Doxygen Html output it would be better to state what the function does via `@brief` statement instead of using the `@returns` statement. If `@brief` is used, Doxygen displays the explanation what the function does already in the summary at the beginning of the Html page. What do you think?

martiniil replied:
Done.

hslusarek reviewed (commented):
Review of the documentation.

I have to admit that I'm not totally sure how much documentation is actually needed. If a function is well named and easily readable, I'm tend to suggest that no additional documentation is needed. But this is just my opinion. I added a comment whenever a documentation was missing. However, I leave it to you, to decide if you actually want to add additional documentation or not.

martiniil commented:
I had a look at the two failing travis jobs. I think one is due to #174.
In the other one the reloading of the controller seems to make problems. The action-client can no longer connect to the server after it. However I could not reproduce it. I have no clue why this happens.

martiniil commented:
@bmagyar This is ready for review.

btw: Me and @hslusarek already spent some time in order to make this PR clean. Now it is up to you. ;)

Commit:
Add extension point in update function to allow additional checks, etc.

- To allow derived classes to perform e.g. additional checks, etc.
  an extension point is added to the the update() function.
- The extension point is an additional protected function which can
  be overwritten by the derived class.
- The stop- and hold-trajectory creation is moved into separate classes,
  to allow derived classes to re-use the stop- and hold-trajectory
  creation code.
- The old desired state and the old time data are now also stored,
  to allow derived classes to perfrom more comprehensive checks, etc.

bmagyar commented:
Note to self, for changelog
```
* Add extension point in update function
* Refactor trajectory building
* Contributors: Pilz GmbH and Co. KG
```

bmagyar commented:
Thanks very much for the work on this guys. As this has been thoroughly reviewed, I took the liberty of splitting it up into smaller, more manageable but logical chunks. I will be merging these parts today as the CI gets progressively happier and will forward port it all to Noetic too. Please note that for future changes, go for Noetic directly.

PS: I think you intentionally blocked my access to your repo which is fair so I couldn't even resolve the conflict using your branch.

bmagyar commented:
https://github.com/ros/rosdistro/pull/24817

https://github.com/ros/rosdistro/pull/24816

martiniil commented:
> Thanks very much for the work on this guys. As this has been thoroughly reviewed, I took the liberty of splitting it up into smaller, more manageable but logical chunks. I will be merging these parts today as the CI gets progressively happier and will forward port it all to Noetic too. Please note that for future changes, go for Noetic directly.
> 
> PS: I think you intentionally blocked my access to your repo which is fair so I couldn't even resolve the conflict using your branch.

Great! The splitting makes sense, thanks!
